modin/core/execution/ray/implementations/pandas_on_ray/partitioning/partition.py

```diff
@@ -93,76 +119,47 @@ def apply(self, func, *args, **kwargs):
+        data, meta, meta_offset = de.exec()
         self._is_debug(log) and log.debug(f"EXIT::Partition.apply::{self._identity}")
-        return self.__constructor__(result, length, width, ip)
+        return self.__constructor__(data, meta=meta, meta_offset=meta_offset)
+
+    @_inherit_docstrings(PandasDataframePartition.add_to_apply_calls)
+    def add_to_apply_calls(self, func, *args, length=None, width=None, **kwargs):
+        return self.__constructor__(
+            data=DeferredExecution(self._data_ref, func, args, kwargs),
+            length=length,
+            width=width,
+        )
 
+    @_inherit_docstrings(PandasDataframePartition.drain_call_queue)
     def drain_call_queue(self):
-        """Execute all operations stored in the call queue on the object wrapped by this partition."""
+        data = self._data_ref
+        if not isinstance(data, DeferredExecution):
+            return data
+
         log = get_logger()
         self._is_debug(log) and log.debug(
             f"ENTER::Partition.drain_call_queue::{self._identity}"
         )
-        if len(self.call_queue) == 0:
-            return
-        data = self._data
-        call_queue = self.call_queue
-        if len(call_queue) > 1:
-            self._is_debug(log) and log.debug(
-                f"SUBMIT::_apply_list_of_funcs::{self._identity}"
-            )
-            (
-                self._data,
-                new_length,
-                new_width,
-                self._ip_cache,
-            ) = _apply_list_of_funcs.remote(data, *deconstruct_call_queue(call_queue))
-        else:
-            # We handle `len(call_queue) == 1` in a different way because
-            # this dramatically improves performance.
-            func, f_args, f_kwargs = call_queue[0]
-            self._is_debug(log) and log.debug(f"SUBMIT::_apply_func::{self._identity}")
-            (
-                self._data,
-                new_length,
-                new_width,
-                self._ip_cache,
-            ) = _apply_func.remote(data, func, *f_args, **f_kwargs)
+        self._data_ref, self._meta, self._meta_offset = data.exec()
+        data.ref_count(-1)
         self._is_debug(log) and log.debug(
             f"EXIT::Partition.drain_call_queue::{self._identity}"
         )
-        self.call_queue = []
-
-        # GH#4732 if we already have evaluated width/length cached as ints,
-        #  don't overwrite that cache with non-evaluated values.
-        if not isinstance(self._length_cache, int):
-            self._length_cache = new_length
-        if not isinstance(self._width_cache, int):
-            self._width_cache = new_width
 
+    @_inherit_docstrings(PandasDataframePartition.wait)
     def wait(self):
-        """Wait completing computations on the object wrapped by the partition."""
         self.drain_call_queue()
-        RayWrapper.wait(self._data)
+        RayWrapper.wait(self._data_ref)
```

Comment:
How should it work if `_data_ref` is `DeferredExecution`?


`TypeError: wait() expected a list of ray.ObjectRef or ray.StreamingObjectRefGenerator`